### ISSUE TITLE
Ability to send message to all Dolores users *right now*

### ISSUE THREAD

**Issue by melodykramer** [18F/dolores-landingham-bot#125]
This is a slightly different use case for Dolores but would immensely help the @18F/comms and @18F/blog teams: the ability to message all Dolores users with the same message at once.

Use cases:

1. Alerting people to new features in Dolores
2. Asking people for information for the all-hands
3. Querying for upcoming blog posts or newsletters

The Comms team needs a way to communicate with each member of 18F on a regular basis. We've been individually DMing people, but that takes time and is not the easiest as 18F scales.

**Comment by jessieay** [comment]
@melodykramer this would be pretty easy to implement.

**Comment by melodykramer** [comment]
I think it would be a huge :+1:  for the @18F/blog team.

**Comment by melodykramer** [comment]
(But I think the admin level should be there too. I'm not sure we'd want everyone to have access to this feature.)

**Comment by awfrancisco** [comment]
So then all the replies go back to Dolores and the comms admin logs in as Dolores and manages all the replies individually?

**Comment by jessieay** [comment]
We aren't currently logging responses sent to Dolores. Does this feature require that? Was thinking it was more of a one way thing. If we need to log responses, that's good to know.

**Comment by melodykramer** [comment]
Yes, we would need a way to log the responses or collect them in some way.

**Comment by gboone** [comment]
I'm not really sure how the sending of messages would work but maybe a "reply-to" feature where replies get sent to a group DM or a channel set by whomever triggers the sending.

**Comment by jessieay** [comment]
@melodykramer if you could get in touch with blog / comms teams and figure out what the most helpful way to display responses would be (group DM or channel) that would be great. 

One limitation here is that I don't know if it will be possible send **certain** Mrs. Landingham replies to certain channels but not others. aka - we could have *all* replies forwarded to an email address, but we can't just have replies to a message sent from comms sent to an email address. Does that make sense? 

Another idea I had for dealing with replies would be to use webhooks, but those are currently only available in channels https://api.slack.com/outgoing-webhooks (not DMs)

So maybe a better way of dealing with this would be to say in the message body something like "DM Melody Kramer if you have any questions"

**Comment by awfrancisco** [comment]
I think a channel would be better than a DM in case the Dolores monitoring team changes or the types of messages sent out change. Not sure how many unsolicited responses Dolores gets at the moment, but might be good to have a channel to drop those into anyway.

**Comment by melodykramer** [comment]
Agree with @awfrancisco

**Comment by melodykramer** [comment]
It would also be useful to send out messages to a subset of Dolores users: i.e. everyone managing a comms project, everyone on the Dev team, etc.

**Comment by melodykramer** [comment]
Just bumping this as it would be a super useful feature for us.

**Comment by ertzeid** [comment]
I'll try to take a look at this this week.

Can an intermediate step for responses be to just include in the message who to follow-up with if people have questions?

**Comment by melodykramer** [comment]
Yes - or a user-group!

**Comment by melodykramer** [comment]
@ertzeid  Like what @awfrancisco said above.

**Comment by jessieay** [comment]
@ertzeid Before we give Dolores these great powers, we want to make sure that only certain users can do this. As part of that effort, last week I added a Users table to the DB so we are saving users and not just using session vars. Let us know if/when you start working on this. When you do, make sure it is a feature only available to admins.

**Comment by jessieay** [comment]
@ertzeid are you still working on this? `admin` stuff is in master now

**Comment by ertzeid** [comment]
It's still a "when I have time" thing, which is ... rarely, these days. If
someone wants to take it, by all means, but I'll keep it on my to-do list
unless I see it reassigned.

**Comment by jessieay** [comment]
# Auction requirements

* At least one high-level (Capybara) test to confirm the following:

> Given that I am an admin
> And and I am logged in
> When I create a new message (with a title and message body)
> Then I should be able to see the message's title and created_at timestamp in ET  in a list of "other messages"
> And I should be able to test a send of that message to one person
> When I click 'Send"
> And I see a js confirm that asks "Are you sure you want to send this message to all Slack users right now?
> And I click "Ok" on the JS confirm
> Then the message should be sent to to all employees in the Dolores system
> And I should see that the message `last_sent_at` datetime was updated (in ET)

* Any other classes/methods that are introduced in order to complete the above feature should have unit tests.

## Other information

This is implementation advice but not required:

This type of message feels significantly different from a regular `scheduled_message` so it may make sense to create a new table called `messages`. The table should have the following fields:

* `created_at` (datetime - required)
* `updated_at` (datetime - required)
* `title` (string - required)
* `body` (text - required)
* `last_sent_at` (datetime - not required)

For viewing messages, there can be a separate section on the sidebar, similar to and below the scheduled messages sidebar:

![screen shot 2016-06-30 at 10 23 46 am](https://cloud.githubusercontent.com/assets/601515/16497567/7a771c7e-3eac-11e6-9398-8116728df7bd.png)

The testing behavior can also work similarly to how testing works for scheduled messages. 

List of messages should look something like this:

![screen shot 2016-06-30 at 10 28 35 am](https://cloud.githubusercontent.com/assets/601515/16497718/3afd7b96-3ead-11e6-91c5-ecb41a9e89c2.png)

We should send the message right away but via delayed job (which we are already using in this codebase) in case there are any failures due to rate limiting.

**Comment by geoffharcourt** [comment]
Confirming receipt of our successful auction bid. Thank you.

**Comment by mrgilman** [comment]
Hi! I'm working on this feature and I have a question.

I noticed that for scheduled messages, there is a `SentScheduledMessage` model that tracks sent messages and, in the `MessageSender` class, also records any errors encountered when sending a message. Does it make sense to create a corresponding `SentMessage` class to go with the new `Message` model? If not, how should error reporting be handled?

**Comment by jessieay** [comment]
Hello @mrgilman  - thanks for asking!

We use `SentScheduledMessages` as a way of knowing who has received a certain messages since we run a rake task to send those every minute (eg: if a user should have received a message at 12 and it is 12:01, we want to check that they did not receive a message before sending it to them)

Recording errors is nice but I hope to make that the job of an error reporting service in the future. As long as we have some type of UI output that notifies the user of any errors, I am not worried about ensuring that we record errors in the DB.

Let me know if you'd like any further clarification! 18F also has some public Slack channels if chat would be easier...

**Comment by mrgilman** [comment]
That should do it for now, thanks @jessieay!
